### PR TITLE
Change synchronized on FileSystemContext to avoid deadlock

### PR DESCRIPTION
**Pull request by elega**
### What changes are proposed in this pull request?

Please outline the changes and how this PR fixes the issue.

### Why are the changes needed?

Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, describe the bug.

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
  1. change in user-facing APIs
  2. addition or removal of property keys
  3. webui

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 1]
```diff
@@ -631,11 +636,14 @@ public synchronized WorkerNetAddress getNodeLocalWorker() throws IOException {
    *
    * @return the info of all block workers eligible for reads and writes
    */
-  public synchronized List<BlockWorkerInfo> getCachedWorkers() throws IOException {
-    if (mWorkerInfoList == null || mWorkerInfoList.isEmpty() || mWorkerRefreshPolicy.attempt()) {
-      mWorkerInfoList = getAllWorkers();
+  public List<BlockWorkerInfo> getCachedWorkers() throws IOException {
+    synchronized (mWorkerInfoList) {
+      if (mWorkerInfoList.get() == null || mWorkerInfoList.get().isEmpty()
+          || mWorkerRefreshPolicy.attempt()) {
```
Comment (jiacheliu3) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java:642]: do we need to worry about the thread safety of this `mWorkerRefreshPolicy`?

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks!

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by yuzhu** [comment]
alluxio-bot, cherry-pick this to branch-2.8 please

**Pull request merged**